### ISSUE TITLE
[Request] Enable 'bots' column by default in server browser

### ISSUE THREAD

**Issue by Splatt581** [ValveSoftware/halflife#2481]
Since some players do not even suspect existence of this hidden useful column, I suggest enabling it by default in all tabs of server browser.
![20190527201559_1](https://user-images.githubusercontent.com/31417200/58436736-ec29d800-80d7-11e9-9d2d-5a5a5a952d65.jpg)

This setting is in `platform/config/InGameDialogConfig.vdf` file. The bots column has label `1_hidden`.

**Comment by sjain882** [comment]
Was this recently introduced? I currently only have access to a 2018 build of HL and I can't find any bot entry in any files of the same name

There isn't even a platform folder in cstrike! D:

**Comment by Splatt581** [comment]
> Was this recently introduced? I currently only have access to a 2018 build of HL and I can't find any bot entry in any files of the same name
> 
> There isn't even a platform folder in cstrike! D:

The `platform` folder is located in the base directory of game - in the `Half-Life` folder, not in the mods folders. The display can also be customized by right-clicking on the line of columns in the server browser.

![bots](https://user-images.githubusercontent.com/31417200/59778939-7117a400-92c8-11e9-9412-1e6f3ec3cc93.jpg)

**Comment by mikela-valve** [comment]
@Splatt581 The Bots column should now show by default in the server browser as of build 8308.

**Comment by basuritashka** [comment]
@mikela-valve Looks like it's not, I haven't touched anything afaik.

![image](https://user-images.githubusercontent.com/33007491/61846931-f1f12f00-ae76-11e9-80c9-46f21ebc3966.png)

**Comment by GiovaniFerraroTrivelli** [comment]
In my case, all columns were enabled after updating the game (not only Bots column, the Server IP too).

**Comment by basuritashka** [comment]
Can confirm this is working but Server IP Address got enabled by default.

**Comment by mikela-valve** [comment]
@GiovaniFerraroTrivelli That makes sense, I think it was due to #2620. The server browser columns are just stored by index, so when I moved the Players column it shifted the index for some of the other columns, so they might be showing when they were previously hidden. 

That should only happen once though, let me know if it resets again for any reason.

**Comment by afwn90cj93201nixr2e1re** [comment]
![изображение](https://user-images.githubusercontent.com/9306996/61898974-8a48eb80-af2b-11e9-8544-fd96c3dd16b0.png)
And why you done changes again? Reorder? For what?

**Comment by afwn90cj93201nixr2e1re** [comment]
![изображение](https://user-images.githubusercontent.com/9306996/61899126-d72cc200-af2b-11e9-9c94-207c2b82ae9b.png)
Just make game column little bit wider and allow only 32 character's. That's not a good solutions for fake server's fix. This is genereal steam master server issue's, not engine.

As i suggested try to make it like: 

> Just make game column little bit wider and allow only 32 character's. 

this will be better than now.

**Comment by afwn90cj93201nixr2e1re** [comment]
@mikela-valve

**Comment by afwn90cj93201nixr2e1re** [comment]
Also as you can see, bot's, vac and close icon's again not centered at all. That's not an ico problem, this is margin problems in code, you can simple fix it.

**Comment by afwn90cj93201nixr2e1re** [comment]
Also seems like serverbrowser do only 1 iterate (this is **bug**) on initializing, and stopped. We should push update all button to make serverbrowser getting new server list.

**Comment by basuritashka** [comment]
#2620 and #2481 were a MUST to fight fake servers that are abusing server browser, and this change is working pretty well. There is no need to change anything right now because it is working flawlessly.

@kisak-valve Sorry for tagging you on this one, but this guy should be banned or some sort of penalization, spamming every thread, non-sense comments and so on.

**Comment by afwn90cj93201nixr2e1re** [comment]
@kisak-valve what bout ban barbie for creating not related to goldsrc engine issues?

He doesnt read anyting and trying iterate my comment's like he want, but he not competent in engine at all.

**Comment by MatiasEsf** [comment]
@afwn90cj93201nixr2e1re You are stupid? (really)

**Comment by SamVanheer** [comment]
First of all the column reordering should be discussed in #2620 

Second, you should all follow the rules of conduct:
https://github.com/ValveSoftware/halflife#conduct

Now, as to the need for a fix in the GoldSource server browser, the server uses this Steam API function to set the game name:
https://github.com/ValveSoftware/source-sdk-2013/blob/0d8dceea4310fde5706b3ce1c70609d72a38efdf/mp/src/public/steam/isteamgameserver.h#L36-L38

As the comment indicates it's used to display servers in the Steam server browser, so changing it will affect *all* games on Steam that use its master server functionality.
Other games will no longer be able to format their description string properly if it's being filtered at that level, so it must be fixed in the GoldSource server browser.

I've noted the existence of #2620 in the Steam server browser in that issue, so it should be all taken care of as far as GoldSource itself goes.

**Comment by CS-PRO1** [comment]
Um, is it me or the bots counter is giving wrong numbers? Sometimes a server is empty and still giving like 10 bots for example. Sometimes server has like 8-10 players and bot count is 16..

**Comment by kisak-valve** [comment]
@afwn90cj93201nixr2e1re, consider this your final warning. You've been given a 1 week timeout for your conduct.

You appear to be more interested in having a fit and calling people out than providing constructive feedback and I've tried to be tolerant under the impression there's a language gap happening here instead of outright trolling, but my patience has worn thin and any further outburst will be viewed dimly.

Please do not force me to ban you outright.

**Comment by norbee16** [comment]
What if the number of seats immediately after the server name? Then that scam thing would not go up and still remain more transparent and beautiful.

**Comment by norbee16** [comment]
What if the number of places immediately after the server name or after the ip address? Then that scam thing would not go up and still remain more transparent and beautifu

**Comment by RauliTop** [comment]
> What if the number of places immediately after the server name or after the ip address? Then that scam thing would not go up and still remain more transparent and beautifu

After the server name they can easily use the same trick like in Game column.
After IP adress will look horrible, too much numbers.

The new visualization it's great. Players and server name are the options most seen.

**Comment by norbee16** [comment]
Yeah, I didn't really think about what I wrote.

**Comment by mikela-valve** [comment]
@CS-PRO1 Are those confirmed good servers that are showing incorrect bot counts? I'll look at how that is reported to see if there's anything weird like it sometimes being uninitialized or something.

Leaving this open for the moment as I'd like to look a bit more into how the column visibility is restored to see if there is anything reasonable that can be done about the wrong columns showing the first time after the update.

**Comment by basuritashka** [comment]
@mikela-valve I am sure it is indeed working but not at all.

`[MAFIACS] Surfitas.cl` is showing 3 bots but their `Game info` is not showing any player/bot connected.
![image](https://user-images.githubusercontent.com/33007491/62012519-59e49580-b155-11e9-9f16-62002aeb8a46.png)

When you are into the server, it will display bots correctly, now 2 bots because one of them left the server.
![image](https://user-images.githubusercontent.com/33007491/62012527-813b6280-b155-11e9-91c2-cf2eee6956ce.png)

Now it will be displayed correctly into the `Game info` panel.
![image](https://user-images.githubusercontent.com/33007491/62012543-a039f480-b155-11e9-84fc-2c36384bf67d.png)

**Comment by CS-PRO1** [comment]
@mikela-valve Well most of the servers picked up by the browser are having this. Like being empty but having bots somehow.. On the servers i usually play there are none having this issue.
![browser](https://user-images.githubusercontent.com/50755959/62012569-b9ac6200-b190-11e9-8cce-209b5a8604c5.png)
Also can you make all tabs having the same column sizing? It's really awkward having to set each individually.. Compare the screenshot above with this one and focus on column sizes and visible columns:
![browser2](https://user-images.githubusercontent.com/50755959/62012600-009a5780-b191-11e9-8646-0ecfbe46e73b.png)

**Comment by uniarch** [comment]
Works for me. Centering vac/bots icons would be good

**Comment by eat1k** [comment]
How it works? I have 31 bots in my server (created with the Potti AMXX plugin) but the server browser doesn't show the num of bots.

**Comment by basuritashka** [comment]
@eat1k There are a lot of ways to implement bots nowadays, probably the way Counter-Strike (and probably other GoldSRC games) is detecting bots hasn't been improved in a while... something must be done there.

**Comment by mikela-valve** [comment]
@eat1k @basuritashka I just checked and it looks like the bot count is just a count of the number of fake clients connected through `pfnCreateFakeClient`, so as noted it would depend on how the bots are implemented whether they'd be detected properly by the current server browser. 

I could add other detection methods but I'd have to know how they can be distinguished from regular clients.

**Comment by basuritashka** [comment]
Probably that check should verify also their authid, AFAIK bots have their `uniqueid` established as `BOT`, server browser should be able to read that unless is not too reliable to check that way.

![image](https://user-images.githubusercontent.com/33007491/62164938-53882200-b2eb-11e9-9b24-0f5d859c8f62.png)

**Comment by eat1k** [comment]
AMXX checks if user is a bot by checking his Auth ID: https://github.com/alliedmodders/amxmodx/blob/2559fcf00aecb19fb6e6f8de846ae37cbcd6d572/modules/dod/dodx/CMisc.h#L141

**Comment by di57inct** [comment]
Bot authids can be spoofed.

**Comment by basuritashka** [comment]
@di57inct The game itself allow `authid` to be spoofed, I don't have any idea which method seems more accurate to deal with how bots are detected. We should add a way to not allow spoofing (encrypting ProtoInfo/TicketInfo/UserInfo) and then `authid` checking can be a good way to detect bots properly.

**Comment by di57inct** [comment]
There's no reliable way to do it since the server can alter every bit of info sent to the client.

**Comment by mikela-valve** [comment]
@basuritashka @eat1k That uniqueid/authid is actually determined the same way that the bot count is for bots. If the client was registered through `pfnCreateFakeClient` it gets the uniqueid/authid of BOT, otherwise it gets the real ID of the client, so any clients showing as BOT through `status` should be getting counted correctly when sending the bot count to Steam.

**Comment by basuritashka** [comment]
@mikela-valve Then for some reason is not behaving correctly or I misunderstood what you said, there are several bots that their `uniqueid` is set as BOT but they are not appearing as a BOT in the Server Browser, unless I am missing something here. As you said, bot counter depends on `pfnCreateFakeClient` and with that BOT authid should be established correctly.

**Comment by basuritashka** [comment]
There is a specific bot (Potti) which relies on Fakemeta to work using `EngFunc_CreateFakeClient` (`pfnCreateFakeClient`) but does not get displayed as a bot by the server browser. Unless I am missing something here.

![image](https://user-images.githubusercontent.com/33007491/62174970-d4eead00-b309-11e9-9592-a7fe285fdfea.png)

**Comment by mikela-valve** [comment]
It looks like there could still be a few things to explain this. The current server information is not reported to Steam all the time, potentially only when players are actually connected to it and playing, it is kept up to date quite consistently in that case. 

When no players are connected it either will only notify Steam on level change or not at all until a player connects, so this could just be Steam showing out of date information from the last time the server reported information like you mentioned [here](https://github.com/ValveSoftware/halflife/issues/2481#issuecomment-515793175).

If you can easily set the number of connected bots from the server console try this:
1. Connect to the server and check the server browser after to see if the number of bots is correct
2. Change the bot count while still connected and see if it updates correctly in the server browser
3. Disconnect from the server
4. Change the bot count and see if it updates in the server browser

It's likely that the first change of bot count will work but the second won't show up in the browser until you connect again to the server.

If it does work like this I might try changing the Steam notifications to still occur when no players are connected just at a much lower rate.

**Comment by basuritashka** [comment]
@mikela-valve That's good to know, everything makes sense now.

So I just managed to add 2 bots to my server, both of them were created successfully when server got initiated.

(Beta Branch Test)
![image](https://user-images.githubusercontent.com/33007491/62175547-1c763880-b30c-11e9-8aa8-a1cd5979f850.png)

They are not being detected by the BOT variable.

I just connected to it and suddenly they are getting recognized as bot as you said.

![image](https://user-images.githubusercontent.com/33007491/62175588-3748ad00-b30c-11e9-8f44-8acf723d06bb.png)

Probably that is the main reason why `BOT Counter` is not updating bot info properly. Your solution would be a good addition.

**Comment by SamVanheer** [comment]
> 
> 
> There is a specific bot (Potti) which relies on Fakemeta to work using `EngFunc_CreateFakeClient` (`pfnCreateFakeClient`) but does not get displayed as a bot by the server browser. Unless I am missing something here.
> 
> ![image](https://user-images.githubusercontent.com/33007491/62174970-d4eead00-b309-11e9-9592-a7fe285fdfea.png)

Just for future reference, `FL_FAKECLIENT` should not be set on `pev->spawnflags`, only on `pev->flags`. I don't think player code uses spawnflags at all but it's good practice to avoid mistakes like this.

From what i can tell the engine tracks fakeclient status in 2 ways: the `FL_FAKECLIENT` flag for game code and `client_t::fakeclient` for engine code. The latter is used to count fake clients to inform Steam.

This information is sent when the server has just finished (re)connecting to Steam, when a map change occurs as well as every frame if maxclients is > 1. I don't see any checks to only send updates if players are connected though.

Note that the engine's server code is responsible for reporting accurate bot counts, so custom engines can be used to bypass this. A modified ReHLDS could do that very easily by calling `SetBotPlayerCount` with incorrect values to spoof that information.

Here's ReHLDS's version:
https://github.com/dreamstalker/rehlds/blob/65c6ce593b5eabf13e92b03352e4b429d0d797b0/rehlds/engine/sv_steam3.cpp#L497-L519

All you have to do here to tell the master server that there are no bots is to hardcode 0, or simply not count fake clients in the loop.

**Comment by afwn90cj93201nixr2e1re** [comment]
@mikela-valve this change's made memory leak on server browser.
Can you fix vgui2?
Coz now when im opening `Find servers` and push Update all button's im getting crash.

As @SamVanheer (a.k.a. Solokiller) said 5-6 years ago, it's because keyvalues in vgui (which serverbrowser.dll use) store something like integer, and not like a string. If you can, fix it. Coz really, i can't update server list without getting crash, i can reproduce it only on Beta branch.

**Comment by basuritashka** [comment]
This should be closed @Splatt581, the main issue got added.

**Comment by mikela-valve** [comment]
@basuritashka I agree, I'm going to close this issue since any fixes required for the counting of bots are better handled in a new issue. 

@afwn90cj93201nixr2e1re if you're getting new crashes in beta I definitely want to check that out before releasing the beta. I'm going to use #2657 to track that. If anyone else is able to reproduce crashes in the server browser as described by @afwn90cj93201nixr2e1re please make a note on #2657.

**Comment by Zetz512** [comment]
I now try to reproduce that problem, but i dont have any crash on refreshing servers from server browser...

on build 6xxx or 7xxx i have that problem but now is all ok

**Comment by oaus** [comment]
@mikela-valve It is not happening on my end, latest beta version
P.S:
![image](https://user-images.githubusercontent.com/43151226/62490636-00005300-b787-11e9-928f-85771f71255c.png)

**Comment by SamVanheer** [comment]
If you're unable to post in an issue it's because the issue creator has blocked you. The creator won't be able to post in your issues either. That may be a bug in Github itself though.

**Comment by afwn90cj93201nixr2e1re** [comment]
Sam unlocked.

**Can you explain** why you changed order?
![изображение](https://user-images.githubusercontent.com/9306996/62496673-bfd2bc00-b7ea-11e9-91b5-17f9faefc7f4.png)
(old)

![изображение](https://user-images.githubusercontent.com/9306996/62496689-ccefab00-b7ea-11e9-85d1-6f486c281527.png)
(new)

We play with a similar order for 10-15 years, some children come (which don’t understand anything at all about finding fake servers), are asked to change and do you change? @mikela-valve 
How to get used to this arrangement now after 10 years? I mean for what this changes, against fake servers or just for Bot columns?


@SamVanheer now you get why im writing here? Im blocked on kid's issues too, but super-moderator don't get it at all...

**Comment by SamVanheer** [comment]
If maintaining column order is important then maybe the column separators could have a border to make any data spilling over obvious at a glance.

**Comment by afwn90cj93201nixr2e1re** [comment]
There ton's of normal ideas without stupid shit as in https://github.com/ValveSoftware/halflife/issues/2620,

Just check it.
We have structure which described at https://developer.valvesoftware.com/wiki/Server_queries#A2S_INFO, when we get response from server(https://github.com/ValveSoftware/halflife/issues/2064#issuecomment-518426948) we just putting it into gameserveritem_t* object, so it's just struct, where every char have **_FIXED_** length 
![изображение](https://user-images.githubusercontent.com/9306996/62499600-5b1c5f00-b7f4-11e9-92a0-cb20546ae925.png)

We can just make columns more wider as i described it in https://github.com/ValveSoftware/halflife/issues/2481#issuecomment-515163172, it's the LOWEST resolution and there no problem with that,
![изображение](https://user-images.githubusercontent.com/9306996/62499738-cc5c1200-b7f4-11e9-9819-dcd4542817af.png)
as you can see,
 but super-moderator prefer ban me, not a kid's which don't get it and start trolling.

**Comment by hashnimmo** [comment]
> afwn90cj93201nixr2e1re commented 16 days ago
> Also as you can see, bot's, vac and close icon's again not centered at all.
I agree with this. So annoying to look at that when you start noticing the little things. Seems this specific update was pushed in a hurry.

**Comment by basuritashka** [comment]
@hashnimmo That was happening since many years ago, and that specifically was not reported. Instead of complaining how updates are being handled start posting some issues to make the game better.

**Comment by hashnimmo** [comment]
@basuritashka So considering it's now being reported, do you mind telling how centering those icons would NOT make the game any better?

If there is no need for complaining, there shouldn't be a so called "Issue" section here in GitHub. It's not like anyone is forcing anyone to do anything here so cool it while the developers are actually free to dump a issue if they don't see any value in it.
